lib/LongVectorLoweringPass.cpp

```diff
@@ -291,44 +325,201 @@ Value *LongVectorLoweringPass::visit(Value *V) {
+  assert(EquivalentValue && "value not lowered");
+
+  Value *ScalarElement = I.getOperand(1);
+  assert(ScalarElement->getType()->isIntegerTy() ||
+         ScalarElement->getType()->isFloatingPointTy());
+
+  ConstantInt *CI = dyn_cast<ConstantInt>(I.getOperand(2));
+  // We'd have to lower this to a store-store-load sequence.
+  assert(CI && "Dynamic indices not supported yet");
+  unsigned Index = CI->getZExtValue();
+
+  IRBuilder<> B(&I);
+  auto *V = B.CreateInsertValue(EquivalentValue, ScalarElement, {Index});
+  registerReplacement(I, *V);
+  return V;
+}
+
+Value *LongVectorLoweringPass::visitShuffleVectorInst(ShuffleVectorInst &I) {
+  assert(isa<FixedVectorType>(I.getType()) &&
+         "shufflevector on scalable vectors is not supported.");
+
+  auto *EquivalentLHS = visitOrSelf(I.getOperand(0));
+  auto *EquivalentRHS = visitOrSelf(I.getOperand(1));
+  auto *EquivalentType = getEquivalentTypeOrSelf(I.getType());
+
+  assert(((EquivalentLHS != I.getOperand(0)) ||
+          (EquivalentRHS != I.getOperand(1)) ||
+          (EquivalentType != I.getType())) &&
+         "nothing to lower");
+
+  IRBuilder<> B(&I);
+
+  // The arguments (LHS and RHS) could be either short-vector or long-vector
+  // types. The latter are already lowered to an aggregate type.
+  //
+  // Extract the scalar at the given index using the appropriate method.
+  auto getScalar = [&B](Value *Vector, unsigned Index) {
+    if (Vector->getType()->isVectorTy()) {
+      return B.CreateExtractElement(Vector, Index);
+    } else {
+      assert(Vector->getType()->isStructTy());
+      return B.CreateExtractValue(Vector, Index);
+    }
+  };
+
+  // The resulting value could be a short or a long vector as well.
+  auto setScalar = [&B](Value *Vector, Value *Scalar, unsigned Index) {
+    if (Vector->getType()->isVectorTy()) {
+      return B.CreateInsertElement(Vector, Scalar, Index);
+    } else {
+      assert(Vector->getType()->isStructTy());
+      return B.CreateInsertValue(Vector, Scalar, Index);
+    }
+  };
+
+  unsigned Arity = I.getShuffleMask().size();
+  auto *ScalarTy = I.getType()->getElementType();
+
+  auto *LHSTy = cast<VectorType>(I.getOperand(0)->getType());
+  assert(!LHSTy->getElementCount().isScalable() && "broken assumption");
+  unsigned LHSArity = LHSTy->getElementCount().getFixedValue();
+
+  // Construct the equivalent shuffled vector, as a struct or a vector.
+  Value *V = UndefValue::get(EquivalentType);
+  for (unsigned i = 0; i < Arity; ++i) {
+    auto Mask = I.getMaskValue(i);
+
+    Value *Scalar = nullptr;
+    if (Mask == -1) {
+      Scalar = UndefValue::get(ScalarTy);
+    } else if (Mask < LHSArity) {
+      Scalar = getScalar(EquivalentLHS, Mask);
+    } else {
+      Scalar = getScalar(EquivalentRHS, Mask - LHSArity);
+    }
+
+    V = setScalar(V, Scalar, i);
+  }
+
+  registerReplacement(I, *V);
+  return V;
+}
+
+bool LongVectorLoweringPass::handlingRequired(Instruction &I) {
```

Comment:
This might be better served taking a User as the parameter for reuse with constants. At some point I assume the lowering will have to break down aggregates that contain long vectors (e.g. arrays and structs) and I think this will be useful at that point.